src/Servers/Kestrel/Core/src/Internal/Http2/Http2HeaderEnumerator.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            _headersEnumerator = default;
+            _trailersEnumerator = default;
+            _genericEnumerator = headers.GetEnumerator();
+            _isTrailers = true;
+
+            _stringValuesEnumerator = default;
+            Current = default;
+        }
+
+        private string GetCurrentKey()
+        {
+            if (_genericEnumerator != null)
+            {
+                return _genericEnumerator.Current.Key;
+            }
+            else if (_isTrailers)
+            {
+                return _trailersEnumerator.Current.Key;
+            }
+            else
+            {
+                return _headersEnumerator.Current.Key;
+            }
+        }
+
+        public bool MoveNext()
+        {
+            if (MoveNextOnStringEnumerator())
+            {
+                return true;
+            }
+
+            if (!TryGetNextStringEnumerator(out _stringValuesEnumerator))
+            {
+                return false;
+            }
+
+            return MoveNextOnStringEnumerator();
+        }
+
+        private bool MoveNextOnStringEnumerator()
+        {
+            var e = _stringValuesEnumerator;
+            var result = e.MoveNext();
+            if (result)
+            {
+                Current = new KeyValuePair<string, string>(GetCurrentKey(), e.Current);
+            }
+            else
+            {
+                Current = default;
+            }
+            _stringValuesEnumerator = e;
+            return result;
+        }
+
+        private bool TryGetNextStringEnumerator(out StringValues.Enumerator enumerator)
+        {
+            if (_genericEnumerator != null)
+            {
+                if (!_genericEnumerator.MoveNext())
+                {
+                    enumerator = default;
+                    return false;
+                }
+                else
+                {
+                    enumerator = _genericEnumerator.Current.Value.GetEnumerator();
+                    return true;
+                }
+            }
+            else if (_isTrailers)
+            {
+                if (!_trailersEnumerator.MoveNext())
+                {
+                    enumerator = default;
+                    return false;
+                }
+                else
+                {
+                    enumerator = _trailersEnumerator.Current.Value.GetEnumerator();
+                    return true;
+                }
+            }
+            else
+            {
+                if (!_headersEnumerator.MoveNext())
+                {
+                    enumerator = default;
+                    return false;
+                }
+                else
+                {
+                    enumerator = _headersEnumerator.Current.Value.GetEnumerator();
+                    return true;
+                }
+            }
+        }
+
+        public KeyValuePair<string, string> Current { get; private set; }
```

Comment:
Why is Current public?

Reply:
This type has the shape of an enumerator, which requires **MoveNext()** and **Current**.

I could make those members internal but it wouldn't change anything. The type is already internal.